lib/manager/npm/post-update/npm.ts

```diff
@@ -31,7 +31,7 @@ export async function generateLockFile(
     if (validRange(npmCompatibility)) {
       installNpm += `@${quote(npmCompatibility)}`;
     }
-    const preCommands = [installNpm];
+    const preCommands = [installNpm, 'hash -d npm'];
```

Comment:
Why?

Reply:
Since `npm` is a symlink, its location is cached by bash. It should be invalidated after `npm` install (at least when I tested on `renovate/node` Docker image).

Reply:
Is this something that changed with npm 7 or could this be treated as a fix independent of npm 7?

Reply:
It is independent of npm 7 but required for this PR to work.

Reply:
Cc @viceice 

npm can also be installed in the Lerna post-update file too. What I don't understand yet:
- Is this caused because the *path* for npm is changed by the install?
- Why didn't this impact us when installing custom versions of node 6 or below?

Reply:
> Is this caused because the path for npm is changed by the install?

After further investigation, yes. It seems the npm bundled with node.js is used and hashed at first, so the path is changed after installing globally:

```
ubuntu@ebe6a367a56e:/usr/src/app$ which npm
/usr/local/node/14.15.0/bin/npm
ubuntu@ebe6a367a56e:/usr/src/app$ type npm
npm is /usr/local/node/14.15.0/bin/npm
ubuntu@ebe6a367a56e:/usr/src/app$ npm -v
6.14.8
ubuntu@ebe6a367a56e:/usr/src/app$ npm i -g npm@7
/home/ubuntu/.npm-global/bin/npm -> /home/ubuntu/.npm-global/lib/node_modules/npm/bin/npm-cli.js
/home/ubuntu/.npm-global/bin/npx -> /home/ubuntu/.npm-global/lib/node_modules/npm/bin/npx-cli.js
+ npm@7.0.10
added 247 packages from 867 contributors in 2.328s
ubuntu@ebe6a367a56e:/usr/src/app$ which npm
/home/ubuntu/.npm-global/bin/npm
ubuntu@ebe6a367a56e:/usr/src/app$ type npm
npm is hashed (/usr/local/node/14.15.0/bin/npm)
ubuntu@ebe6a367a56e:/usr/src/app$ npm -v
6.14.8
ubuntu@ebe6a367a56e:/usr/src/app$ hash -d npm
ubuntu@ebe6a367a56e:/usr/src/app$ npm -v
7.0.10
```

> Why didn't this impact us when installing custom versions of node 6 or below?

It did, but it seems nobody have realized.

Reply:
👀 So we silendly used the default global npm version silently ? so we need to check  where this can happen on other places of renovate.

Are am right, that the hash is only valid for current bash instance?

Reply:
@viceice 
> Are am right, that the hash is only valid for current bash instance?

Yes, it is.